Automattic/jetpack
# Connection: Limit repeated failed attempts to update remote package versions

This PR adds caching login in `Package_Version_Tracker` so that when the remote request to WPCOM in order to update the tracked DNA package versions fails, it will only attempt to update it again once an hour.
The reason we are doing this is that the local package versions will be actually updated only when the corresponding remote request succeeds. However, if it doesn't (aka WPCOM returns a non-200 response code for any reason) it will repeatedly attempt this request. Hence, adding a caching login there, will ensure that in these edge cases, clients (remote Jetpack sites) will limit these attempts to an hourly rate.

#### Changes proposed in this Pull Request:

* Introduces a new transient: `jetpack_failed_update_remote_package_versions` set to expire hourly
* Sets the above transient to the current datetime in case the request to WPCOM to update the package versions fails (Note here that we are not actually interested to the set value when checking, rather it's existence)
* Checks the existence of the aforementioned transient before attempting to update the remote package versions and if set, returns early.

#### Jetpack product discussion
1200561894344934-as-1200777997736934/f

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
- Modify your sandbox to always return a non-200 status code when hitting the `jetpack-package-versions` endpoint
- Spin up a new JN site from the **master branch** and sandbox it
- Set up Jetpack Connection (site-only should be enough)
- Delete the local tracked versions: `wp option delete jetpack_package_versions`
- Tail `debug.log` and note that the remote requests to `jetpack-package-versions` occur repeatedly on each page load.
- Using **Jetpack Beta Tester**, switch to this branch
- Tail `debug.log` and ensure the remote request to `jetpack-package-versions` occurred only once.
